compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -998,8 +998,14 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             | ty::Alias(..)
             | ty::Param(..)
             | ty::Bound(..)
-            | ty::Error(_) => {}
-            ty::Infer(_) => {
+            | ty::Error(_)
+            | ty::Infer(
+                ty::InferTy::IntVar(_)
+                | ty::InferTy::FloatVar(_)
+                | ty::InferTy::FreshIntTy(_)
+                | ty::InferTy::FreshFloatTy(_),
+            ) => {}
+            ty::Infer(ty::InferTy::TyVar(_) | ty::InferTy::FreshTy(_)) => {
                 candidates.ambiguous = true;
```

Comment:
This is the only case I can think of that could still cause issues, I just hope this was already always ambiguous anyway :crossed_fingers:.

Reply:
I tried to call a method on a ty infer var but that caused errors on typeck about not being able to do that

Reply:
Yeah, that's what I ran into as well, when trying to reproduce it w/o floats.

AFAIK you can't get it to pick a trait based on method name alone, without something like `Default::default().method()` or `<_>::method` but those give up too early because of how ambiguous they are, only int/float inference vars go farther (because they represent an unresolved choice in a small finite set of types etc.).

The fuzzier ideas I've had would involve guiding inference, but `_: FnPtr` will never succeed early enough to taint inference, and blocking inference would likely require associated types etc. - and the moment you pin down a trait, `FnPtr` can only affect it through a blanket impl of your choice, which would also apply to all the existing types too (before `FnPtr` was added).